Hosting: Allow all sites to see the Hosting section in development.

kwight opened the pull request:
Allow all sites to see the Hosting Config section in their menu; this requires enabling a feature flag by query param or environment variable. This will prepare us for presenting an upgrade path to those noticing the feature and wanting it for their currently ineligible site.

#### Testing instructions

* Start this PR in your local dev environment with `ENABLE_FEATURES=hosting/all-sites npm start`.
* Select a VIP site and a Jetpack site; verify neither has Manage > Hosting Configuration in the sidebar, nor access to the page directly by URL (`https://calypso.localhost:3000/hosting-config/:domain`).
* Select a Simple site that has no plan or the Premium plan.
* Verify Manage > Hosting Configuration appears in the sidebar, and that the page is accessible.
* Select a site Simple site on the Business plan; verify Hosting features are accessible.
* Select an Atomic site; verify Hosting features are accessible.

Fixes #36489

gwwar commented on config/development.json:
```diff
@@ -66,6 +66,7 @@
 		"help": true,
 		"help/courses": true,
 		"hosting": true,
+		"hosting/all-sites": true,
```
Maybe keep this off, I think we confused other devs last time. we can enable with ?flags=hosting/all-sites or starting with `ENABLE_FEATURES=hosting/all-sites npm start`

kwight replied:
ce07082f575cac16bd6ba4cc93069cf2db53c7f6

gwwar commented on client/state/selectors/can-site-view-atomic-hosting.js:
```diff
@@ -20,9 +21,17 @@ export default function canSiteViewAtomicHosting( state ) {
 	}
 
-	return canCurrentUser( state, siteId, 'manage_options' );
+	if ( false !== isJetpackSite( siteId ) || false !== isVipSite( siteId ) ) {
```
Don't forget `state`! ✨This will cause Calypso to not load.

`isJetpackSite( state, siteId )`
`isVipSite( state, siteId )`

kwight replied:
Oops, lol. ccda6fc72c29b69e7a2ead779a2d1a8b760ad14a

mmtr commented on client/state/selectors/can-site-view-atomic-hosting.js:
```diff
@@ -20,9 +23,21 @@ export default function canSiteViewAtomicHosting( state ) {
 	}
 
-	return canCurrentUser( state, siteId, 'manage_options' );
+	if ( false !== isJetpackSite( state, siteId ) && true !== isSiteAtomic( state, siteId ) ) {
```
I had to read the line a couple of times to understand this condition 😅. Is there any reason why this cannot be simplified as below?
```suggestion
	if ( isJetpackSite( state, siteId ) && ! isSiteAtomic( state, siteId ) ) {
```

kwight replied:
> Is there any reason why this cannot be simplified as below?

These selectors can return `null` if the site being a Jetpack site is undetermined. With `null` being falsey, an `isJetpackSite` check will evaluate `false` if it's unknown if the site is a Jetpack site, and continue instead of bailing. I only wanted to allow continuing if we're certain the site is not a Jetpack site, hence the awkward direct comparison of `false`. You're right, it's horrible looking and mind-bending, but it ensures that only sites we _know_ are Atomic sites will continue (same thing with VIP).

jeraldjuice replied:
What if you added a condition for `! site` at the top and didn't rely on looking for `null`s here? that might be more declarative and ensure nothing slips through.

```
const site = getSite( state, siteId );
if ( ! site ) {
     return false;
}
```

jeraldjuice replied:
Or could also just add an extra condition with lodash `isNull`

mmtr commented on client/state/selectors/can-site-view-atomic-hosting.js:
```diff
@@ -20,9 +23,21 @@ export default function canSiteViewAtomicHosting( state ) {
+		return false;
+	}
+
+	if ( false !== isVipSite( state, siteId ) ) {
```
Same as below here:
```suggestion
	if ( isVipSite( state, siteId ) ) {
```

gwwar commented:
Breaking this out a bit with comments for intent, you mean the following right?

```jsx
// early return on non-admin role
if ( ! canCurrentUser( state, siteId, 'manage_options' ) ) {
    return false;
}

// early return if VIP
if ( isVipSite( state, siteId ) ) {
    return false;
}

// early return if Jetpack standalone
if ( isJetpackSite( state, siteId ) && ! isSiteAtomic( state, siteId ) {
   //not sure if the second selector is still correct
   return false;
}

// okay for everyone else
if ( isEnabled( 'hosting/all-sites' ) ) {
	return true;
}

//otherwise check by plan
return isSiteOnAtomicPlan( state, siteId );
```

kwight commented:
```js
if ( isVipSite( state, siteId ) ) {
    return false;
}
```

This will continue rather than bailing if the selector cannot determine if the site is VIP or not; the selector will return `null`, which is falsey, and allow execution to continue. Same with `isJetpackSite`.

gwwar commented:
From changes in D20666-code I think it's pretty safe for us to trust the is_vip option.

kwight commented:
> From changes in D20666-code I think it's pretty safe for us to trust the is_vip option.

Agreed, but that assumes it's made it into state and the site object is properly loaded by the selectors (`getRawSite` returns a site).

kwight commented:
In the latest updates, I've added a dedicated selector for WP.org Jetpack (non-Atomic) sites), and removed the yoda formatting. I think this is as clear and safe as it can be; we'll defensively bail if any site selector returns `null`.

But if people think this is still too much, than say the word and I'll commit the clearest (but less defensive) version in @gwwar 's [comment](https://github.com/Automattic/wp-calypso/pull/38155#issuecomment-562302263).

kwight commented:
I like @jeraldjuice 's suggestion the most; defensively verify a valid site object early, so as not to worry about handling `null` values later (meaning the more straight-forward truthy checks can be used).
